WordPress/gutenberg
# Fluid typography: add font size constraints

Resolves https://github.com/WordPress/gutenberg/issues/44957

WordPress/wordpress-develop patch:

- https://github.com/WordPress/wordpress-develop/pull/3489

Cherry picked into wp/6.1 branch in this PR:

- https://github.com/WordPress/gutenberg/pull/45133

## What?

This PR adds default minimum font size constraint of 14px/0.875rem/0.875em so that a fluid font's minimum font size does not become smaller than the constraint.

❗ For user-defined font sizes that are < 14px we will use the font-size itself as the minimum value. This is so that changes to these custom values are reflected in the editor and frontend. We trust the user here to insert a sensible value

## Why and how?
To ensure that minimum font sizes, whether user-defined or calculated by `gutenberg_get_typography_font_size_value` are not less than 14px/0.875rem/0.875em.

This is to protect readability and accessibility.

## Testing Instructions
In theme.json, add some very small font sizes (<14px or .875rem) to the presets' `fluid.min`.

Example:

```json
				{
					"size": "2.8rem",
					"fluid": {
						"min": "0.3rem",
						"max": "30px"
					},
					"slug": "neutrale",
					"name": "Neutrale"
				},
```

Add a post and apply the preset.

Check that these font sizes are never less than 14px or .875rem in the editor and frontend.

Next add a custom font size to an element, either in theme.json or via the editor.

If your custom font size is <14px or .875rem, the value of the custom font size will be static.

Otherwise the clamped value's minimum font size should be never be less than 14px or .875rem.

<details>

<summary>Example theme.json</summary>

```json
{
	"version": 2,
	"settings": {
		"typography": {
			"fluid": true,
			"fontSizes": [
				{
					"size": ".9rem",
					"slug": "piccolino",
					"name": "Piccolino"
				},
				{
					"size": "2rem",
					"fluid": {
						"min": "1.8rem",
						"max": "40px"
					},
					"slug": "in-mezzo",
					"name": "In Mezzo"
				},
				{
					"size": "2.8rem",
					"fluid": {
						"min": "0.3rem",
						"max": "30px"
					},
					"slug": "neutrale",
					"name": "Neutrale"
				},
				{
					"size": "4.75rem",
					"slug": "grandone",
					"name": "Grandone"
				}
			]
		}
	},
	"styles": {
		"typography": {
			"fontSize": "12px"
		},
		"elements": {
			"h1": {
				"typography": {
					"fontSize": "4rem"
				}
			}
		},
		"blocks": {
			"core/post-date": {
				"typography": {
					"fontSize": "16px"
				}
			}
		}
	}
}

```

</details>

<details>

<summary>Example HTML</summary>

```html

<h2>Default font size in theme.json style (not clamped because &lt; 14px):</h2>



<p>Lorem ipsum dolor sit amet. A eaque harum in aspernatur vero qui sapiente asperiores quo harum exercitationem qui quasi neque et laborum veritatis. Et sint inventore rem aliquam tempora nam ratione quaerat.</p>



<h2>Corrected font size in preset (won't be smaller than 14px):</h2>



<p class="has-piccolino-font-size">Lorem ipsum dolor sit amet. A eaque harum in aspernatur vero qui sapiente asperiores quo harum exercitationem qui quasi neque et laborum veritatis. Et sint inventore rem aliquam tempora nam ratione quaerat.</p>



<h2>Corrected fluid.min font size in preset (won't be smaller than 14px):</h2>



<p class="has-neutrale-font-size">Lorem ipsum dolor sit amet. A eaque harum in aspernatur vero qui sapiente asperiores quo harum exercitationem qui quasi neque et laborum veritatis. Et sint inventore rem aliquam tempora nam ratione quaerat.</p>



<h2>Small custom font size (not clamped because &lt; 14px):</h2>



<p style="font-size:12px">Lorem ipsum dolor sit amet. A eaque harum in aspernatur vero qui sapiente asperiores quo harum exercitationem qui quasi neque et laborum veritatis. Et sint inventore rem aliquam tempora nam ratione quaerat.</p>

```

</details>
## Screenshots or screencast 

https://user-images.githubusercontent.com/6458278/196073775-1459f362-cbc1-4e36-81aa-e46660730ad1.mp4
